### DISCUSSION TITLE
Mesh

### DISCUSSION THREAD

**Discussion by kasia129** [capytaine/capytaine#114] [Q&A] [answered]
Hi, is there any recommended way of preparing meshes for capytaine? The bodies I'm going to run my analyses on will be quite simple but a bit more complex than spheres or cylinders for which you provide predefined methods (assembly of truncated cones, most probably).

**Comment by rgcoe** [comment]
I like to use [`pygmsh`](https://github.com/nschloe/pygmsh). Here's an example from one of tests in `capytaine`: 

https://github.com/mancellin/capytaine/blob/cb22cd2b455217dcbbcc19abaea3b52a35b42f92/pytest/test_io_meshes.py#L103-L128

**Reply by kasia129** [reply]
Thanks a lot! Since I'll be working with rather simple geometries defined by a set of points, I'm inclining towards a simpler solution, as explained below. If that fails, I may come back to the solution you suggested above.

**Comment by kasia129** [comment]
I'm currently trying what seems to be the simplest approach: *AxialSymmetricMesh.from_profile()*, providing the profile as an array. To test it, I'm trying to mesh a simple vertical cylinder of length 1 and radius 1, placed directly below the free surface, in two ways:

1) using *bodies.predefined.cylinders.VerticalCylinder()*:

```buoy = cpt.bodies.predefined.cylinders.VerticalCylinder(length=1.0, radius=1.0, center=(0, 0, -.5), ntheta=40, name="my buoy")```

When solving the problem, I am getting the following warning:

>  The mesh of the body axisymmetric_mesh is not inside the domain.

I'm able to solve this issue by changing center to `(0,0,-.501)`.

2) using *AxialSymmetricMesh.from_profile()*:

```
vertex0 = [ 0,   0,  -1]
vertex1 = [-1,   0,  -1]
vertex2 = [-1,   0,   0]
vertex3 = [ 0,   0,   0]

shape = np.array([vertex0, vertex1, vertex2, vertex3])

buoy = cpt.FloatingBody(cpt.AxialSymmetricMesh.from_profile(profile=shape, nphi=40))
```
Here I am getting the same warning, and solve it be changing `[-1,   0,   0]` to `[-1,   0,   -0.01]` and `[0,   0,   0]` to `[0,   0,   -0.01]` . Why putting the cylinder right at the free surface causes errors? Is there any other way to solve this rather than "artificially" shifting the floater?

I would expect the two methods of meshing to be equivalent, what would allow me verification of method 2. However, running the same problem on these two meshes gives different results (in terms of added mass). Why are these two methods not equivalent? I would appreciate any hints.

**Reply by mancellin** [reply]
> Why putting the cylinder right at the free surface causes errors? Is there any other way to solve this rather than "artificially" shifting the floater?

At the moment, the solver requires all mesh faces to be stricly below the free surface. You can manually remove faces right on the free surface with `keep_immersed_part()`. Removing the (0, 0, 0) point in `from_profile()` might also avoid generating faces on the free surface.

> Why are these two methods not equivalent?

A possible difference between them is the number and shape of mesh faces. Did you try looking at the generated meshes with the 3D viewer?

**Reply by kasia129** [reply]
Thanks a lot for the suggestions. I tried both solutions:

adding `buoy.keep_immersed_part()` resulted in the warning, as before:

> WARNING:capytaine.bem.problems_and_results:The mesh of the body axisymmetric_mesh is not inside the domain.
> Check the position of the free_surface and the sea_bottom
> or use body.keep_immersed_part() to clip the mesh.

Removing the (0,0,0) point I was able to run the problem without the warning, however, I got a lot different result than expected (i.e., than the one obtained with the mesh created with the *.VerticalCylinder* method).

As for plotting, I am having trouble installing VTK, so I am not able to plot the mesh. Unless there is another way?

**Reply by mancellin** [reply]
> adding buoy.keep_immersed_part() resulted in the warning, as before

That's weird. Possibly a bug in the mesh clipper.

> As for plotting, I am having trouble installing VTK, so I am not able to plot the mesh. Unless there is another way?

If you have matplotlib installed, there is `buoy.show_matplotlib()` method to display the mesh.

**Comment by kasia129** [comment, marked as answer]
Thanks. I've placed buoy.keep_immersed_part() right after defining the buoy object:

```
buoy = cpt.FloatingBody(cpt.AxialSymmetricMesh.from_profile(profile=shape, nphi=nphi))
buoy.keep_immersed_part()
```

Can you confirm that is correct. 

The plot of the mesh looks good, no red flags there I think:

![image](https://user-images.githubusercontent.com/76471057/143412690-61a81b61-ff27-4818-a5b2-3405c404e093.png)

After clipping (buoy.keep_immersed_part()) the plot looks exactly the same: should the panels on the top of the cylinder be removed from the plot?

**Reply by mancellin** [reply]
> Can you confirm that is correct.

Looks good to me.

> should the panels on the top of the cylinder be removed from the plot?

Yes, since they are not wet, these panels should be removed from the mesh. In principle, `keep_immersed_part()` should remove all panels with `z >= 0`. There can be an issue, for instance when the top panels are actually `z = -1e-12`, or something like that.

**Reply by kasia129** [reply]
Thanks. I managed to get the keep_immersed_part() function work by first extending the body surface above the free surface (i.e., above `z=0`), and then clipping it at `z=0`. It seems that for clipping to work, the extreme points of the body must be located above the free surface `z=0`, and not just at the free surface. As you suggested, that seems to be a matter of precision.
